Update demo app to include style switcher/setter

tallytalwar opened the pull request:
- Choose between CDN bundled scene files
- Or specify a URL for a scene to be loaded

A custom ArrayAdapter was required to specify a custom filter to include a string outside of the initial filter list.

Closes #1362

sarahlensing commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -88,6 +189,63 @@ public void onLowMemory() {
 
+    @Override public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+        final ArrayList<SceneUpdate> sceneUpdates = new ArrayList<>(1);
+        final String apiKey = "vector-tiles-tyHL4AY";
```
remove key and replace with value that is read from gradle.properties?

1. add to `~/.gradle/gradle.properties`:
```
tangramDemoApiKey=the_api_key
```

2. add to demo app `build.gradle`:
```
def DEMO_API_KEY = '"' + tangramDemoApiKey + '"';

buildTypes {
    debug {
      debuggable true
      jniDebuggable true
      buildConfigField 'String', 'API_KEY', DEMO_API_KEY
    }
  }
```

3. update this line to: `final String apiKey = BuildConfig.API_KEY;`

tallytalwar replied:
This is good.

We have an issue to track the work for the same for all platform demos. https://github.com/tangrams/tangram-es/issues/1452

tallytalwar replied:
Refactored a bit to at least not have same scene API setting 3 times in the code.

sarahlensing commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -88,6 +189,63 @@ public void onLowMemory() {
+        final String apiKey = "vector-tiles-tyHL4AY";
+        sceneUpdates.add(new SceneUpdate("global.sdk_mapzen_api_key", apiKey));
+
+        switch (position) {
```
since you do the same thing for all index values < scenes.length, you could refactor this to something like:

```
if (position < scenes.length) {
   sceneString.setText(scenes[position]);
   map.loadSceneFile(scenes[position], sceneUpdates);
} else {
   sceneString.setText(null);
   sceneString.requestFocus();
   InputMethodManager in = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
   in.showSoftInput(sceneString, 0);
}
```

the else clause adds text clearing and shows the keyboard so the user can enter the custom scene location (if you don't want this behavior feel free to ignore though :) )

tallytalwar replied:
Works great. Addressed.

sarahlensing commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -88,6 +189,63 @@ public void onLowMemory() {
+        final String apiKey = "vector-tiles-tyHL4AY";
+        sceneUpdates.add(new SceneUpdate("global.sdk_mapzen_api_key", apiKey));
+        spinner.setSelection(7);
+        map.loadSceneFile(sceneString.getText().toString(),sceneUpdates);
```
if you replace `onItemClick` with something like this:

```
int selectedIndex = 7; // custom
String sceneText = sceneString.getText().toString();
for (int i = 0; i < scenes.length; i++) {
    if (sceneText.equals(scenes[i])) {
        selectedIndex = i;
        break;
    }
}
spinner.setSelection(selectedIndex);
        
sceneString.clearFocus();
InputMethodManager in = (InputMethodManager) getSystemService(
Context.INPUT_METHOD_SERVICE);
in.hideSoftInputFromWindow(sceneString.getWindowToken(), 0);
```

it will add some polish by dismissing the keyboard on selection and detect the corresponding spinner item selected and update it to be the selected spinner item (right now if you click a style within the scenes list it wont update the spinner).

tallytalwar replied:
Works great. Addressed.

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,18 +39,90 @@
+        "https://mapzen.com/carto/zinc-style/zinc-style.zip"
+    };
+
+    AutoCompleteTextView sceneString;
```
Addressed.

matteblair commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,36 +40,130 @@
         view = (MapView)findViewById(R.id.map);
         view.onCreate(savedInstanceState);
+        autoCompleteView.setText(scenes[0]);
         view.getMapAsync(this, "scene.yaml", sceneUpdates);
```
This call should use the current text in the `AutoCompleteView` instead of a constant.

tallytalwar replied:
Addressed.

matteblair commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -88,6 +191,42 @@ public void onLowMemory() {
     }
 
+    @Override public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+        int selectedIndex = 7; // custom
```
This smells like trouble. `onItemClick` and `onItemSelected` seem overly complicated and dependent on data that's defined elsewhere. As in my earlier comment, I'd suggest removing the `Spinner` element entirely.

tallytalwar replied:
Implicitly handled by removing spinner.

matteblair commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -88,6 +191,42 @@ public void onLowMemory() {
+        InputMethodManager in = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
+        in.hideSoftInputFromWindow(autoCompleteView.getWindowToken(), 0);
+
+        map.loadSceneFile(autoCompleteView.getText().toString(),sceneUpdates);
```
Loading a scene file synchronously from a network location on the UI thread is ... _no bueno_. Android could very well kill our app or produce the famous "Application Not Responding" message if this takes too long. However, there does seem to be a problem with calling `MapView$getMapAsync` more than once per `MapView` ... so either we live with an unresponsive UI for now or we fix `getMapAsync`.

tallytalwar replied:
Agreed. I would suggest implementing `loadSceneFileAsync` call in our `MapController` to address this issue. But outside the scope of this PR and can be addressed in a separate PR.

tallytalwar commented on platforms/android/demo/build.gradle:
```diff
@@ -25,4 +25,5 @@ android {
 
 dependencies {
   compile project(path: ':tangram')
+    compile 'com.android.support:design:25.3.1'
```
My Android Studio settings did this. Addressed.

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,36 +40,130 @@
+    private final String apiKey = "vector-tiles-tyHL4AY";
+    private ArrayList<SceneUpdate> sceneUpdates = new ArrayList<>();
+
+    /* Override ArrayAdaptor to include any text */
```
Addressed.

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,36 +40,130 @@
+    private ArrayList<SceneUpdate> sceneUpdates = new ArrayList<>();
+
+    /* Override ArrayAdaptor to include any text */
+    private class CustomArrayAdapter extends ArrayAdapter {
```
Make sense, now done in a separate class `DemoSceneManager`. The most apt name I could come for `CustomArrayAdapter` is `SceneAutoCompleteArrayAdapter`.

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,36 +40,130 @@
+            }
+        };
+
+        CustomArrayAdapter(Context context, int resource, List<String> objects) {
```
I think you can do that. I get an `java.lang.UnsupportedOperationException`. Checked a couple of stackoverflows and they say because one can not `clear` the contents as is being done in the custom filter here. Of course if I am missing something java(y) then I am not sure.

sarahlensing replied:
Yep, thats right, when you use `String[]`, the adapter creates an immutable list

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -31,36 +40,130 @@
+
+    AutoCompleteTextView autoCompleteView;
+    CustomArrayAdapter customAdapter;
+    Spinner spinner;
```
Addressed. no spinner now. Make sense.

tallytalwar commented:
@sarahlensing can you again give this a look, post the changes. Thanks

sarahlensing commented:
@tallytalwar LGTM

tallytalwar commented:
Updated the previous commit with couple of tiny improvements:
1. Switched to use `more-label` variants of house style from cdn
2. AutoCompleteTextView now highlights the whole text on getting focus. Makes it easy for the user to get rid of the entire text.

matteblair commented:
The last issue I'm trying to track down in this is that when the `MainActivity` is re-created due to rotating the screen, the text in the `AutoCompleteTextView` is preserved from the last instance, but the scene itself always reverts to the original "scene.yaml".

What's weird is that `MainActivity$onCreate` gets the scene URL from the current text of the `AutoCompleteTextView`, but the preserved text isn't restored in that view until _after_ `onCreate`!

matteblair commented:
The first quick "solution" that I thought of was just to make the `AutoCompleteTextView` _not_ preserve its text, but this behavior seems to be an irrevocable part of `EditText` and its subclasses (https://developer.android.com/reference/android/widget/TextView.html#setFreezesText(boolean)).

tallytalwar commented on platforms/android/demo/src/main/java/com/mapzen/tangram/android/MainActivity.java:
```diff
@@ -76,8 +76,16 @@ public void loadSceneCallback(String scene) {
+    }
+
+    @Override
+    public void onPostCreate(Bundle savedInstanceState) {
```
👍
